### ISSUE TITLE
XGrammar's Logits Modification Affects Model Capability in Tool-Calling Scenarios

### ISSUE THREAD

**Issue by Csrayz** [mlc-ai/xgrammar#581]
Our team has been integrating XGrammar into our vLLM to improve tool-calling accuracy in production. While XGrammar has successfully reduced failure rates caused by JSON schema violations, we noticed that it appears to affect the model's overall capability during tool calls.

~~We believe one potential cause is that XGrammar alters the output logits distribution. Specifically, after setting the probabilities of illegal tokens to zero, the remaining token distribution becomes sharper — producing an effect similar to lowering the temperature. This concentration of probability mass may inadvertently bias the model's outputs.~~

We're wondering if there is a way to preserve the original logits distribution as closely as possible after masking illegal tokens. Below is a simple illustration:
```
>>> a = torch.tensor([[0.2, 0.7, 0.1]])
>>> torch.softmax(a, -1)
tensor([[0.2814, 0.4640, 0.2546]])

>>> a[0][0] = -torch.inf  # mask illegal token
>>> torch.softmax(a, -1)
tensor([[0.0000, 0.6457, 0.3543]])
```

Would you consider this a known concern, or is there an existing approach to mitigate this effect?
